### ISSUE TITLE
Any IMAP folder (other than Inbox): Run filter "Folder Name contains" on email arrival and/or periodically

### ISSUE THREAD

**Issue by 03302024** [RealRaven2000/FiltaQuilla#318]
I added FiltaQuilla extension and set it up as the following and it does not perform any action:

![392666837-b1e20095-fc2b-4201-813e-88f64100076f](https://github.com/user-attachments/assets/ab2938fa-9d02-4aaf-8b59-a296a87b10b9)

The issue that this would resolve for me is this: The IMAP email server, which is out of my hands, erroneously puts good emails in the Bulk folder. In Thunderbird, I need to manually copy all new emails from IMAP Bulk folder to Local Inbox folder. It would be nice if this action could be done automatically, on email arrival and/or periodically.

**Comment by CherryDT** [comment]
+1

The help page even mentions that the "Folder Name" filter would be useful in case of incoming filters putting stuff into folders, but it seems it's not very useful if the filters don't even run on folders other than Inbox. How could it be achieved?

**Comment by RealRaven2000** [comment]
> +1
> 
> The help page even mentions that the "Folder Name" filter would be useful in case of incoming filters putting stuff into folders, but it seems it's not very useful if the filters don't even run on folders other than Inbox. How could it be achieved?

...by installing an older version of Thunderbird and using the inherited folder properties feature. Thunderbird 68 should probably work. 

at the moment I have no clear way of implementing this in modern versions of tb 115 or 128. automatic filter triggering everywhere on demand would be nice though.  I will ask about it in the next Dev Call ( next Thursday). meme you could also add a feature request to the Thunderbird Bugzilla...

**Comment by Betterbird** [comment]
> using the inherited folder properties feature

What's that?

**Comment by RealRaven2000** [comment]
> > using the inherited folder properties feature
> 
> What's that?

It was a feature in the version for Thunderbird 60 which overlayed the folder properties dialog. It added an option to execute the filters automatically on this folder, too. It got lost when I converted the Add-on for Thunderbird 68 or 102 I cannot remember. I am not sure whether it is a good feature to be provided by an Add-on, really feel it should be the app that does it.

looking at the comments now (folderPropsOverlay.js) - from the Tb60 branch:

```js
    // setup UI for the "applyIncomingFilters" inherited property, but only for
    // imap non-inbox folders.

```

in filtaquilla.js:


```js 
  // inherited property object
  let applyIncomingFilters = {
    defaultValue: function defaultValue(aFolder) {
      return false;
    },
    name: filtaquillaStrings.GetStringFromName("filtaquilla.applyIncomingFilters"),
    accesskey: filtaquillaStrings.GetStringFromName("filtaquilla.applyIncomingFilters.accesskey"),
    property: "applyIncomingFilters",
    hidefor: "nntp,none,pop3,rss" // That is, this is only valid for imap.
  };
```

from what I can see in comm-esr128, there is an `m_applyIncomingFilters` property, which is implemented in `[nsImapMailFolder.cpp](https://searchfox.org/comm-esr128/source/mailnews/imap/src/nsImapMailFolder.cpp)` only

**Comment by Betterbird** [comment]
An inherited property is just one where the getter also checks the parent recursively.

All that's needed is to run `setStringProperty("applyIncomingFilters", "true");` on the folder. And yes, `m_applyIncomingFilters` is an object member that is set from the folder property.

**Comment by RealRaven2000** [comment]
> An inherited property is just one where the getter also checks the parent recursively.
> 

ok, but does it mean it is only useful for child folders of _Inbox_? (which has the flag hardcoded, I guess) - I need to learn more why exactly that is there, to make a folder and all its children react to filters?

> All that's needed is to run `setStringProperty("applyIncomingFilters", "true");` on the folder. And yes, `m_applyIncomingFilters` is an object member that is set from the folder property.

Even if I don't modify folderProperties dialog we could add a webext interface into the folder tree context menu to set / reset it. 

But obviously I need to do a little more research on the folder name method to test this. I think it would be quite useful to extend quickFilters as well (which has a global search method for filters that target folders) to be able to list all filters (quickFilters extends the search box of the message filters dialog) that apply to a specific folder. 

The problem, if it does not have the folder name search term, does this mean all filters will autorun on the folder? or does it only run the filters with that search term? The later would be better as it would enable the user to target specific folders with specific filters. 

Can somebody raise a feature request [in FiltaQuilla](https://github.com/RealRaven2000/FiltaQuilla/issues) to support setting the `applyIncomingFilters` flag on any folder. And why not in POP3, how is it even dependent on the mail protocol? Was just a experimental thing they tested out in Thunderbird ages ago?

**Comment by Betterbird** [comment]
You have a feature request right here, why raise another one?

Note that the add-on "Folder Flags" modifies the panel by adding a tab, so adding a checkbox should be easier.

The folder tree context menu would be a very ugly place to put it.

POP only delivers to the inbox, so it makes no sense there. For that reason, the underlying TB code only checks the flag for IMAP folders. Kent implemented that stuff in TB to serve his add-on, see: https://bugzilla.mozilla.org/show_bug.cgi?id=257415

**Comment by RealRaven2000** [comment]
So I need to look into "folder name contains" operator it may be more useful if it would target a specific folder, as a special flag also needs to be added to the folder. Just a mere string check on the folder name may be too unspecific. I might make the feature dependent on a quickFilters Pro license to fund it, but we will see about that.

**Comment by RealRaven2000** [comment]
So here is a test version for the feature:

[filtaquilla-4.3.2pre11.zip](https://github.com/user-attachments/files/18671734/filtaquilla-4.3.2pre11.zip)

I added the option "run filter on incoming mails" to the folder properties (general tab) - but only for IMAP folders:

![Image](https://github.com/user-attachments/assets/2a53962c-1252-46a8-8075-228bba1dc1a6)

it seems to work for newly arriving mails in the folder, and seems to be able to match the folder name. The only disadvantage is that literally all filters for the account will be applied to all incoming mails unless we modify the filters to exclude the folder. I think it would be much more useful if only filters that had a folder verification would actually work in a particular folder and the other ones would be ignored, it is not really feasible to modify all filters to exclude  certain folders?

**Comment by 03302024** [comment]
Thank you for the test version. In addition to "Run filters on incomming mails", is it possible to have "Run filters periodically every x minutes", as well?

**Comment by RealRaven2000** [comment]
> Thank you for the test version. In addition to "Run filters on incomming mails", is it possible to have "Run filters periodically every x minutes", as well?

that's a good question! It's worth testing it out.

I think it _should_ work but I can find out on the next developer meeting. If not, we might be able to add that as a feature request to Thunderbird Core. If it's a low effort fix we might submit the patch ourselves. (the main thing is that no new UI is required so there is no problems with localization / string freezes etc. And it wouldn't affect "ordinary users" because they have no way of setting the `applyIncomingFilters` property other than FiltaQuilla.

If it was technically difficult, then I should probably look into supporting something like periodic filtering with a UI from QuickFolders, because that already has a data structure (and UI) in place to do special stuff with special folders:

![Image](https://github.com/user-attachments/assets/da382af9-dbb4-4799-aa9b-243da850ddec)

UI is a bit messy right now because it needs to be fully transitioned from xhtml to html, but it does serve a purpose; I could imagine adding a filter interval such as "run periodic filters every ... minutes".

**Comment by 03302024** [comment]
> I could imagine adding a filter interval such as "run periodic filters every ... minutes".

Thank you. I also think the filter has to priodically trigger the downloading process for the specific IMAP folder. Else, it will just periodically check for email arrival in that specific folder, which never happens.

**Comment by RealRaven2000** [comment]
Here is a new version, which is also compatible with the latest beta 136.0b2:

[filtaquilla-4.3.2pre26.zip](https://github.com/user-attachments/files/18849124/filtaquilla-4.3.2pre26.zip)

this includes the changes in  #331 because Mozilla removed `ChromeUtils.import` and `XPCOMUtils.defineLazyGetter` in version 136.

**Comment by RealRaven2000** [comment]
I believe that **quickFilters Pro** is the correct place to add the UI for allowing automatic filtering on a folder, not QuickFolders, so I added a reference on [this quickFilters issue](https://github.com/RealRaven2000/quickFilters/issues/6) and pinned it to be done as one of the next features

**Comment by RealRaven2000** [comment]
Request from another user:

> It would be great if all the filters could use the default for all filters such as inbox, **while one filter is designated to run against a specific folder like deleted**. 

So at the moment _all filters_ (of the account) are run, which is not only a drain on performance but also probably not desired. I believe the filtering here should be restricted to a sublist of filters that contain the "folder name contains" search term at least once.

from OP:

> Thank you. I also think the filter has to periodically trigger the downloading process for the specific IMAP folder. Else, it will just periodically check for email arrival in that specific folder, which never happens.

ok let's not lose this requirement either. My preference would be to implement this as a [quickFilters Pro](https://addons.thunderbird.net/en-US/thunderbird/addon/quickfilters/) advanced feature because there will be considerable work and risk of support if I give it away for free! It is not an easy to document (and possibly dangerous) feature, that should be used by responsible users only.

**Comment by mpatca** [comment]
I'm not sure I've understood the discussion. I would therefore just like to mention another use case, if it can already be implemented with what has been planned so far, all the better: 

if all IMAP folders and their contents could be included just like the inbox folder, and in particular if the X-day filter function could also be used, there would be the possibility of realising a fine-grained automatic archiving function with filters. In my opinion, this would be a great advantage.

**Comment by RealRaven2000** [comment]
> if all IMAP folders and their contents could be included just like the inbox folder, and in particular if the X-day filter function could also be used, there would be the possibility of realising a fine-grained automatic archiving function with filters. In my opinion, this would be a great advantage.

With the feature included [in the comment above](https://github.com/RealRaven2000/FiltaQuilla/issues/318#issuecomment-2666267469) it is theoretically possible, but  you have to explicitly enable filtering (read: running ALL filters) on each folder individually, to be set through "folder properties":

![Image](https://github.com/user-attachments/assets/4962c59f-1a22-48d7-99f4-afe195ec8576)

**Comment by mpatca** [comment]
I understand the procedure and see the need to include each folder individually as an advantage for the reasons mentioned. 

Unfortunately, it doesn't work automatically for me, whereas the filters work when triggered manually. Could the reason be that I have set mail.server.default.applyIncomingFilters as a string to true (Betterbird 128.8.0esr-bb23, Windows 10)?

**Comment by RealRaven2000** [comment]
> Could the reason be that I have set mail.server.default.applyIncomingFilters as a string to true (Betterbird 128.8.0esr-bb23, Windows 10)?

where? couldn't find that string in about:config, so I do not know where you set it?

**Comment by mpatca** [comment]
This is said to be [a barely documented function](https://superuser.com/questions/544480/how-to-apply-automatic-message-filters-to-subfolders-too) (FiltaQuilla is also mentioned in this context), which should have [worked in at least some versions of Thunderbird](https://bugzilla.mozilla.org/show_bug.cgi?id=428574).

**Comment by MurphySunset** [comment]
I'm also running into this need! Thanks, RealRaven2000, for the great work. I'll be following this issue closely.

**Comment by RealRaven2000** [comment]
Implemented in 5.0 - published 05/04/2025

**Comment by RealRaven2000** [comment]
@Betterbird  5.0 was rejected by the review team, because they didn't want us to add a checkbox to the folder properties dialog. I might just add a switch to only allow this in Betterbird, if they would allow it there?

i then plan to later add a web extension compatible interface in the form of a context menu on the folder tree.

**Comment by Betterbird** [comment]
This doesn't require an answer, does it? I'm surprised that TB add-an review is this strict when on the other hand they don't care about add-ons which obliterate the host application. We're glad not to host important add-ons at ATN any more.

**Comment by RealRaven2000** [comment]
there is stiff push back at the moment to me even adding the code (even if it doesn't affect Thunderbird at all). If it gets too much I will just pull the feature for now until I have the wx-compatible interface). I guess I will have to rip the inheritedProperties stuff out.

**Comment by CherryDT** [comment]
Yes please keep it working in Betterbird! 😅 

For anyone reading this later on - the version 5.0 that was pulled which does have this feature and also other fixes compared to 4.3.2-pre11 posted before is here: https://www.betterbird.eu/downloads/Addons-fixed-for-BB/FiltaQuilla-5.0.xpi

**Comment by RealRaven2000** [comment]
> Yes please keep it working in Betterbird! 😅
> 

I will add an API compliant user interface (folder tree context menu) later, which will b available for both Betterbird and Thunderbird users.

**Comment by Betterbird** [comment]
There is no 5.0 version available on the releases page. We built one since people keep asking, for example here:
https://www.reddit.com/r/Betterbird/comments/1kiz0pe/why_dont_filters_run_automatically_on_the_junk/

BTW, filtaquilla-background.html is missing from build.manifest, that appears to be a bug since it's referenced in manifest.json (Issue #348).

![Image](https://github.com/user-attachments/assets/3f07fca6-f877-478b-8a1c-01cfec61519e)

https://www.betterbird.eu/downloads/Addons-fixed-for-BB/FiltaQuilla-5.0.xpi

**Comment by RealRaven2000** [comment]
> There is no 5.0 version available on the releases page. We built one since people keep asking, for example here:

yes, it was removed by the review team. I promise I will build an API compliant UI (context menu), but I had to fix saving attachments first, which is going into v5.2. Since I am totally interested in running all filters automatically in any folder of the user's choice, if that's at all possible - I am considering adding the UI as a feature in quickFilters Pro.

**Comment by mat926** [comment]
Hello I'm facing the same issue on version 5.2. I created some filters to directly send the mail sent by some senders to the trash (pishing), the problem is that, since the mail server classifies it as Junk (Outlook), this mail goes directly to the Junk folder and not to the Inbox, does this prevent the filter from working automatically? Because every time I have to do it manually and it doesn't do it automatically.

**Comment by RealRaven2000** [comment]
> Hello I'm facing the same issue on version 5.2. I created some filters to directly send the mail sent by some senders to the trash (pishing), the problem is that, since the mail server classifies it as Junk (Outlook), this mail goes directly to the Junk folder and not to the Inbox, does this prevent the filter from working automatically? Because every time I have to do it manually and it doesn't do it automatically.

yeah that's a bummer, we have no influence on the actions that the server takes (unless you can whitelist the address there). Not sure whether automatical filters run on whats incoming in Junk. Would need to ask somebody from Thunderbird staff..

**Comment by Betterbird** [comment]
Maybe we should add this to BB 140 after all since we're in the process of finalising new strings.

EDIT: We'll be shipping this in BB 140. Implementation time < 2h incl. localisation into 16 locales:
https://github.com/Betterbird/thunderbird-patches/commit/31200fdfb2a20925467e9a65d2d454fb3dc1805f
https://github.com/Betterbird/thunderbird-patches/commit/23f3260730be2620c56f6e74f31a3f9967be43bd

**Comment by RealRaven2000** [comment]
> Maybe we should add this to BB 140 after all since we're in the process of finalizing new strings.

that's a great idea! I can still add an interface from FiltaQuilla (via the folder context menu) later to activate the option for less fortunate (Thunderbird users). Custom search terms can always be enabled individually anyway depending on mail client. I was just busy with making attachments and sounds work again, so that's what's going into the next version for FiltaQuilla.

**Comment by RealRaven2000** [comment]
> Maybe we should add this to BB 140 after all since we're in the process of finalising new strings.
> 
> EDIT: We'll be shipping this in BB 140. Implementation time < 2h incl. localisation into 16 locales: [Betterbird/thunderbird-patches@31200fd](https://github.com/Betterbird/thunderbird-patches/commit/31200fdfb2a20925467e9a65d2d454fb3dc1805f) [Betterbird/thunderbird-patches@23f3260](https://github.com/Betterbird/thunderbird-patches/commit/23f3260730be2620c56f6e74f31a3f9967be43bd)

It didn't take long to implement in Thunderbird either, but I will play by the rules and add it to quickFilters Pro as a perk. Then people got a choice to either use "native support" in betterbird or install quickFilters to manage it in Tb.

I uploaded a test version here:

https://github.com/RealRaven2000/quickFilters/issues/305#issuecomment-3016012062

UI - works within the permitted confines of wx development:

![Image](https://github.com/user-attachments/assets/44cc91e5-ec62-448e-b220-1b5516f36ce0)

Anybody on Thunderbird (or Betterbird) feel free to test if it works as expected!

**Comment by RealRaven2000** [comment]
Implemented in 6.7 - Published 23/07/2025
